### PR TITLE
Intro nitllvm, a (work in progress) compiler to LLVM IR

### PR DESCRIPTION
**Pull request by xymus**
This PR adds to code written by @Oaujdi Brini for INF7641 with some modifications.

What nitllvm does:
* Generate class structures and function signatures in LLVM. As stated in the scope of the current project. it does not generate method implementation code.
* Generated LLVM code compilable by clang.
* Can compile `base_simple` but not `hello_world`. (Probably an unresolved type...)
* It is a good base for further development.

I've modified it so:
* The Nit code is prettier and faster, mainly by using `Array[String]` and `join` instead of appends.
* The output is testable.
* Tweaked the LLVM code so it compiles. They are temporary hacks but still ensure that the generated code is valid. One of these hack breaks the generated VFT, I've made a note in the code and it should easy to fix for someone who knows LLVM IR.

### CODE REVIEW

File: src/nitllvm.nit [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+						params.add "%class_{mpropdef.mclassdef.name}*"
+						llvm_function.add "@{index}(%class_{mclass.name}* %this"
+
+						for param in mpropdef.msignature.mparameters do
+							llvm_function.add ",%class_{param.mtype.name} %{param.name}"
+							params.add "%class_{param.mtype.name}"
+						end
+						class_vft_entry.add params.join(",")
+
+						llvm_function.add """) nounwind \{
```
Comment (Morriar) [inline src/nitllvm.nit]: Is the `\{` a Nit escape? If yes, you don't need it in a triple quoted string.

**Comment by Morriar** [comment]
+1

File: src/nitllvm.nit [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+				if mproperty isa MMethod then
+
+					var mpropdefs = mproperty.lookup_all_definitions(mainmodule, mclass.mclass_type)
+					for mpropdef in mpropdefs do
+
+						var class_vft_entry = ["\t"]
+
+						var index = "{mpropdef.mclassdef.name}_{mpropdef.name}"
+						var llvm_function = ["define "]
+						if	mpropdef.msignature.return_mtype != null then
```
Comment (R4PaSs) [inline src/nitllvm.nit]: The tab after the if here kind of bugs me

File: src/nitllvm.nit [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+		for class_name, llvm_code in llvm_attributes do ll.write llvm_code
+
+		ll.write "\n; Functions\n"
+		for llvm_code in llvm_functions.values do ll.write llvm_code
+
+		ll.write "\n; Classes VFTs\n"
+		for llvm_code in llvm_classes_vfts.values do ll.write llvm_code
+
+		ll.write """
+define i32 @main(i32 %argc, i8** %argv) nounwind \{
```
Comment (R4PaSs) [inline src/nitllvm.nit]: Same remark as @Morriar expressed earlier, since we're in a triple-quoted, there's no need for this escape sequence

**Comment by R4PaSs** [comment]
The IR is surprisingly elegant, I was expecting something along the lines of assembly, I'm surprised !

:shipit: !

**Comment by xymus** [comment]
Rerolled for comments.

**Comment by Morriar** [comment]
+1

**Comment by jpages** [comment]
Very, very nice :)

+1

**Comment by privat** [comment]
ca serait cool que pour https://github.com/xymus/nit/commit/e1e8a2f0242978529283920e3b5ff5df6778bb55 l'auteur et le signed-off-by soit l'auteur et pas xymus (le committer reste xymus)

De mémoire, on peut modifier l'auteur du dernier commit avec
~~~sh
$ git commit --amend --author "John Doe <jdoe@example.com>"
~~~

**Comment by xymus** [comment]
Rerolled for @privat comments.

**Comment by BlackMinou** [comment]
+1 ! 
black-minou-mini

File: src/nitllvm.nit [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+					var mpropdefs = mproperty.lookup_all_definitions(mainmodule, mclass.mclass_type)
+					for mpropdef in mpropdefs do
+
+						var class_vft_entry = ["\t"]
+
+						var index = "{mpropdef.mclassdef.name}_{mpropdef.name}"
+						var llvm_function = ["define "]
+						if mpropdef.msignature.return_mtype != null then
+							llvm_function.add "%{mpropdef.msignature.return_mtype.name} "
```
Comment (privat) [inline src/nitllvm.nit]: 7th indentation level. nice.

File: src/nitllvm.nit [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+						params.add "%class_{mpropdef.mclassdef.name}*"
+						llvm_function.add "@{index}(%class_{mclass.name}* %this"
+
+						for param in mpropdef.msignature.mparameters do
+							llvm_function.add ",%class_{param.mtype.name} %{param.name}"
+							params.add "%class_{param.mtype.name}"
+						end
+						class_vft_entry.add params.join(",")
+
+						llvm_function.add """) nounwind {
```
Comment (privat) [inline src/nitllvm.nit]: nounwind?

File: src/nitllvm.nit [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+	redef fun process_mainmodule(mainmodule, mmodules)
+	do
+		# Prepare LLVM code in memory
+		var llvm_classes_vfts = new HashMap[String, String]
+		var llvm_functions = new HashMap[String, String]
+		var llvm_attributes = new HashMap[String, String]
+
+		for mclass in mainmodule.flatten_mclass_hierarchy do
+			var class_vft = ["%class_{mclass.name}_vft = type \{\n"]
```
Comment (privat) [inline src/nitllvm.nit]: should use `cname` instead of `name`

File: src/nitllvm.nit [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+			var class_vft_entries = new Array[String]
+			var class_attributes = ["\n%class_{mclass.name} = type \{%class_{mclass.name}_vft*"] # TODO
+
+			var mproperties = mclass.all_mproperties(mainmodule, private_visibility)
+			for mproperty in mproperties do
```
Comment (privat) [inline src/nitllvm.nit]: The order of `mproperties` does not offer any guarantee. Especially no guarantee about any position invariance with regard to super-classes.

File: src/nitllvm.nit [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+				if mproperty isa MMethod then
+
+					var mpropdefs = mproperty.lookup_all_definitions(mainmodule, mclass.mclass_type)
+					for mpropdef in mpropdefs do
```
Comment (privat) [inline src/nitllvm.nit]: I do not understand why all (re-)definitions of a method should be considered

File: src/nitllvm.nit [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+						var class_vft_entry = ["\t"]
+
+						var index = "{mpropdef.mclassdef.name}_{mpropdef.name}"
```
Comment (privat) [inline src/nitllvm.nit]: just use `mpropdef.cname`

File: src/nitllvm.nit [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+		# Write LLVM code to file
+		var ll: Writer
+		var llvm_output_file = toolcontext.opt_output.value
+		if llvm_output_file != null then
+			ll = llvm_output_file.to_path.open_wo
+		else ll = sys.stdout
+
+		ll.write "\n; Class Structures\n"
+		for class_name, llvm_code in llvm_attributes do ll.write llvm_code
```
Comment (privat) [inline src/nitllvm.nit]: the two following `for` use `.values`, why not this one?

File: src/nitllvm.nit [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+# process options
+var toolcontext = new ToolContext
+toolcontext.process_options(args)
+var arguments = toolcontext.option_context.rest
+
+# build model
+var model = new Model
+var mbuilder = new ModelBuilder(model, toolcontext)
+var mmodules = mbuilder.parse_full(arguments)
```
Comment (privat) [inline src/nitllvm.nit]: engines usually do not want to use `parse_full` but just `parse`

File: src/nitllvm.nit (hunk 5 above)
Comment (MehdiAit) [inline src/nitllvm.nit]: @privat this attribut (or annotation) is used to indicate that the function never raises an exception !

**Comment by privat** [comment]
I will just close the PR as I thing that, with the current state of the code, it will cause some pain to maintain because it is a client of non-stable API (libs and classes in src) without any real benefit.
The PR and the code will be archived by github anyway so it will be not completely lost.

**Pull request closed without merging**